### ISSUE TITLE
Auto backup single-player worlds

### ISSUE THREAD

**Issue by Ashwin-droid** [PolyMC/PolyMC#1076]
### Goal

Backup single-player worlds automatically on exit. Except not if it is unchanged.

### Motivation

People make a lot of progress in Single-player worlds including world records, youtube videos, hardcore series, etc but a small mistake can be well very bad. So I think adding something like snapshots every time you exit the game. Delete old snapshots (should be customizable) I would set it to like 30 times and permanent snapshots every month/year to keep track of your progress in that month/year

### Specification

People make a lot of progress in Single-player worlds including world records, youtube videos, hardcore series, etc but a small mistake can be well very bad. So I think adding something like temporary snapshots and permanent snapshots.
Temporary snapshots:- every time you exit the game. Delete old snapshots (should be customizable) I would set it to like 30 times
Permanent snapshots:- every month/year to keep track of your progress in that month/year
Use case for permanent snapshots:- say that I start a world to build a maze or a game which I would improve daily with new obstacles, etc. A permanent snapshot would keep a track of it and the ability to preview the snapshot

### Drawbacks

Might use up a lot of storage if the size is huge
So a few possible solutions 

* Using git VCS
* disable the option
* make less frequent backups 
* offload to cloud

### Unresolved Questions

🤷‍♂️  everybody needs this

### Alternatives Considered

Doing git myself with task scheduler and cmd

### This suggestion is unique

- [x] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

**Comment by ZekeZDev** [comment]
I agree this is something we need and should probably be a default. 

I am not to sure what the best approach would be especially considering worlds can get rather huge when modded. Not to mention if we would want to backup other things such as configs and some mod data (outside of world folder). 

Cloud backups is an interesting idea however I dont know how feasable it would be especially on limited connections or with storage caps.

**Comment by RaptaG** [comment]
If added, please make it togglable

**Comment by flowln** [comment]
IMO this is more of a job for a mod, rather than us. AFAIK there's already plenty of mods that do just that, for various versions / mod loaders. One example that comes to mind is [AromaBackup](https://www.curseforge.com/minecraft/mc-mods/aromabackup). [Textile backup](https://modrinth.com/mod/textile_backup) also seems interesting, although i've never used it myself.

I would say there'd be a point in doing instance-wide backups, such as for easy backup when updating mods and modpacks, but for worlds specifically, i think the currently available options work just fine.

**Comment by Scrumplex** [comment]
I also think, that this should just be handled by a mod ingame.

**Comment by DioEgizio** [comment]
i agree that this should be handled by a mod in the game. There are already many mods that allow you to do backups, for example the ones that flow cited and [FTB Backups](https://www.curseforge.com/minecraft/mc-mods/ftb-backups-forge)/[FTB Backups 2](https://www.curseforge.com/minecraft/mc-mods/ftb-backups-2) come to my mind.

**Comment by Ashwin-droid** [comment]
I kindly do not agree with this because 
- firstly, speed runs wouldn't be accepted.
- secondly, you need to mod Minecraft for that ( you cannot play vanilla).
- I am more of a fabric guy than a forge one.
- it is also version dependent so no snapshots ( I like the april fools snapshots to play occasionally)

Edit :- these are the reasons for which I am doing git myself instead of modding it

**Comment by DioEgizio** [comment]
There are some fabric mods to do it though, and speedrunning is a niche use case

**Comment by nift4** [comment]
Can't the post-run commands be used for this?

**Comment by OZoneGuy** [comment]
Adding to the discussion.

Another use case is resetting, read deleting and redownloading a mod pack, an instance and wanting to save the world.

You can copy the world manually, but would prefer a built-in way to manage this.

We can do incremental backup, to avoid having multiple copies of the same world. Maybe use a tool like `robocopy`? I don't think implementing this is feasible.
